src/webapi/date.rs

```diff
@@ -1,25 +1,1166 @@
-use std::marker::PhantomData;
-use webcore::try_from::TryInto;
+use std::collections::HashMap;
+use webcore::try_from::{TryFrom, TryInto};
+use webcore::value::{Value, Reference};
 
-/// [(JavaScript docs)](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Date)
-#[derive(Debug)]
-pub struct Date {
-    dummy: PhantomData< () >
-}
-
-impl Date {
-    /// The Date.now() method returns the number of milliseconds elapsed since 1 January 1970 00:00:00 UTC.
+/// A date object
+///
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Date)
+pub trait IDate: AsRef< Reference > + TryFrom< Value >  {
+    /// The Date::now() method returns the number of milliseconds elapsed since 1 January 1970 00:00:00 UTC.
     ///
     /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/now)
-    pub fn now() -> f64 {
+    #[inline]
+    fn now() -> f64 {
+        js!( return Date.now(); ).try_into().unwrap()
+    }
+
+    /// The Date::utc() method accepts the same parameters as the longest form of the constructor,
+    /// and returns the number of milliseconds in a Date object since January 1, 1970, 00:00:00,
+    /// universal time
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/utc)
+    #[inline]
+    fn utc( year: i32, _month: Option< i32 >, _day: Option< i32 >, _hour: Option< i32 >, _minutes: Option< i32 >, _seconds: Option< i32 >, _milliseconds: Option< i32 >) -> i64 {
```

Comment:
Yep, this is better!

However I still think the `Option`s are unnecessary; taking everything as `i32` would be simpler with very little downsides.

Reply:
Sounds fair enough - if we're not shooting for exact parity with the browser's UTC usage, then I can make the change. in progress

Reply:
We're definitely not aiming for absolute 100% coverage of the APIs, **especially** for APIs which would be super awkward like this one and aren't really essential. (I know this reasoning is kind-of wishy-washy, but it's hard to define an absolute rule to which we can adhere to besides using common sense.)